opentech/apply/projects/views.py

```diff
@@ -54,32 +55,35 @@ def get_context_data(self, **kwargs):
         context['contracts'] = contracts
         return context
 
-    def get_contracts(self):
+    def get_contracts(self, project):
         """
         Get approved AND signed contracts for the current project
 
-        When an unsigned and unapproved contract is newer than the signed and
-        approved ones we add that to the top of the list.
+        When an unsigned OR (signed AND unapproved) contract is newer than the
+        signed and approved ones we add that to the top of the list.
         """
-        project = self.get_object()
-
         contracts = (project.contracts.select_related('approver')
                                       .order_by('-created_at'))
 
         if not contracts.exists():
             return
 
         latest_contract = contracts.first()
-        if not (latest_contract.is_signed and latest_contract.approver):
-            # remove unsigned contracts EXCEPT this PK
-            exclude_pks = (contracts.filter(is_signed=False)
-                                    .exclude(pk=latest_contract.pk)
-                                    .values_list('pk', flat=True))
-            contracts = contracts.exclude(pk__in=exclude_pks)
-        else:
-            contracts = contracts.exclude(is_signed=False)
-
-        return contracts
+
+        if latest_contract.is_signed and latest_contract.approver:
+            return contracts.exclude(is_signed=False)
+
+        # At this point we know that latest_contract is the "pending" contract
+        # we want to display at the top of the list, however we need to remove
+        # all the older pending contracts from the list.
+        #
+        # We do that by getting the PKs for all the contracts which are (signed
+        # AND approved), adding the latest_contract's PK and returning the
+        # filter of those.
+        ready_pks = list(contracts.filter(is_signed=True, approver__isnull=False)
+                                  .values_list('pk', flat=True))
+
+        return contracts.filter(pk__in=[latest_contract.pk, *ready_pks])
```

Comment:
This all feels a bit much for what its trying to do. 

I think the methods should be: (method names w.i.p)
```python
def get_context_data(self, **kwargs):
    <cut>
    latest_unapproved_contact = self.get_the_contract_we_are_trying_to_get_approved()
    contracts = project.contracts.select_related('approver').order_by('-created_at')
    <cut>

def get_the_contract_we_are_trying_to_get_approved(self):
    latest = project.contracts.select_related('approver').order_by('-created_at').first()
    if not latest:
        return None
    if latest.approver:
         return None
    return latest
```

Reply:
Fixed in 9f825fc2